server/src/main/java/org/apache/iotdb/db/mpp/plan/planner/distribution/ExchangeNodeAdder.java

```diff
@@ -412,6 +414,10 @@ public boolean equals(Object o) {
   }
 
   private PlanNode processMultiChildNode(MultiChildProcessNode node, NodeGroupContext context) {
+    if (hasVirtualSourceNode(node)) {
```

Comment:
You can put whether having VirtualSourceNode in `NodeGroupContext` instead of recursively judge here. 
Actually for some queries that have many PlanNodes, this `hasVirtualSourceNode` is not efficient beacuse it will iterate all the planodes. So it's better to put this information in `NodeGroupContext` since you've already know it outside.